src/capture/capture_video.cpp

```diff
@@ -355,36 +355,63 @@ static void create_avi_file(const uint16_t width, const uint16_t height,
 	video.audio.bytes_written   = 0;
 }
 
-// Reads the image's pixels from the emulated DOS video RAM (which is always
-// little-endian) and passes them in the same byte-order to the encoder.
+// Performs some transforms on the passed down rendered image to make sure
+// we're capturing the raw output, then passes the result in the same
+// byte-order to the encoder. Endianness varies per pixel format (see
+// PixelFormat in video.h for details); the ZMBV encoder handles all that
+// detail.
 //
-// Ideally this could just be one memcpy, but many DOS video modes need their
-// pixels either doubled horizontally and/or vertically to produce the correct
-// on-screen size. So extra copies are needed in both these cases.
+// We always write non-double-scanned and non-pixel-doubled frames in raw
+// video capture mode :
 //
-static void compress_frame(const RenderedImage& image)
+// - Pixel-doubling is not a problem as that's always performed as a
+//   post-processing step; it's never "baked-into" the rendered image.
+//   We just need to completely ignore the `double_width` flag.
+//
+// - Double-scanning can be either "baked-in" or performed in post-processing.
+//   Ignoring the `double_height` flag takes care of the post-processing
+//   variety, but for "baked-in" double-scanning, we need to skip every second
+//   row.
```

Comment:
That's just how it works 🤷🏻 Maybe doing the "fake doubling" via the scaler post-VGA-render is more efficient (probably it is), but not sure how much.

That VGA code is fraught with danger, it's super tricky and dangerous to modify it. At some point @kcgen attempted to "true double scan" CGA on VGA, but it didn't go that well... 😅 VGA and EGA modes on VGA are "true double scanned", anything else on VGA that needs double scanning (e.g., CGA, and VESA modes that need double scanning) are "fake double-scanned".

I am planning to document the VGA rendering pipeline as I've spent a few hundred hours this year understanding / refactoring / modifying that code 😅 Then we'll have an accurate description of the current state of things. Trying to "normalise" things (e.g. by getting rid of fake double scanning) will be the job for someone who has the appetite to sink yet another 20-50-100+ hours into this.

And wait, we also have pixel-doubling... Both true and fake... 😬

Reply:
>  VGA and EGA modes on VGA are "true double scanned", anything else on VGA that needs double scanning (e.g., CGA, and VESA modes that need double scanning) are "fake double-scanned".

The 320X200 resolution that Doom outputs (which I believe is a standard VGA mode) is a real 320X200 image.  I saw that from stepping through this PR with a debugger and while working on ffmpeg.  You're saying with "true double scanning", it should be 320X400?  That is not the case currently.

>  (Well, but at least I can say with confidence now I fully understand all that, for all video modes on all emulated adapters.)

At least one of us does.

>  I am planning to document the VGA rendering pipeline

That would be great.

Reply:
> The 320X200 resolution that Doom outputs (which I believe is a standard VGA mode) is a real 320X200 image. I saw that from stepping through this PR with a debugger and while working on ffmpeg. You're saying with "true double scanning", it should be 320X400? That is not the case currently.

Well, it depends on a few factors. Read this function, especially the top comment. The TL;DR is that yes, if you use the `sharp` shader` or the `texturenb`, 320x200 is rendered as 320x200 because you'd get absolutely no difference in the final output if we rendered it as 320x400 (either "true" or "fake").

https://github.com/dosbox-staging/dosbox-staging/blob/792386135985e9b8e55f42546c32fe7e07aba529/src/gui/render.cpp#L554-L599

But if you enable CRT shaders, the 200 vs 400 line difference is absolutely crucial. 320x200 is really 640x400 on VGA. VGA cannot "do" 200-line modes, not really. Low-res modes are upscaled by 2x _in hardware (!)_ behind the scenes. So with CRT shaders, we must have 400 scanlines; 200 scanlines would VGA output look like some arcade cabinet or home computer 😄

Reply:
In real hardware, there are two stages:

1. Game -> writes pixel data to VGA memory
2. Video card draws scan lines to the screen

What you saw, @weirddan455 is at step 1.

It's at step 2 where the video card will do line double for the lower resolution VGA modes.

DOSBox has to emulate both steps. The final destination for us, the "screen", is SDL or OpenGL.  So DOSBox does the line-doubling just like the video card does.

(You might wonder.. _"well, why not let the GPU double those lines - it would be faster in hardware"_) But we lose accuracy; have the actual final "screen pixels" let John do the very accurate VGA glshading, and it even helps modes like `output = texture` look pretty close to the real thing, too, because having twice as many rows tightens up the interpolation.

Reply:
> In real hardware, there are two stages:
> 
>     1. Game writes pixel data to VGA memory (and sets mode and timing registers)
> 
>     2. Video card draws scan lines to the screen

Yeah, it's good to spell that out. I should start doing that because to me it's obvious, but I guess not spelling it out might confuse people 😅